### PR TITLE
[dynamo] Reuse AsyncCollectiveTensor-traced graph for the resolved Tensor input

### LINKED ISSUE #187073: torch.compile recompiles when input alternates between AsyncCollectiveTensor and Tensor
## Summary

When a compiled region's input alternates between `AsyncCollectiveTensor` (ACT)
and a resolved `Tensor` across steps, Dynamo guards on the tensor class and
recompiles each time the class changes. With FSDP + TP this happens during the
first several steps (all-gathered weights are natively `AsyncCollectiveTensor`,
but get resolved to a plain `Tensor` once awaited before the compiled region by
prefetch/overlap scheduling), forcing users to raise
`torch._dynamo.config.cache_size_limit` well above the default of 8.

The all-gather output is fundamentally an `AsyncCollectiveTensor`; the plain
`Tensor` case only arises when the ACT is awaited before the compiled region.
So the direction that matters here is **trace with ACT, then provide a resolved
`Tensor` at runtime** — which the runtime wrapper already handles by unwrapping
ACT. Relaxing the class guard for *this* direction should be safe, and would
avoid the wasted recompiles.

(Note the reverse direction — trace with `Tensor`, then provide an un-awaited
ACT at runtime — is *not* safe to relax, since the traced graph may skip the
await; this repro deliberately does not exercise that direction.)

## Repro

```python
# TORCH_LOGS=recompiles torchrun --nproc_per_node=2 repro.py
import torch
import torch.distributed as dist
from torch.distributed._functional_collectives import all_gather_tensor

dist.init_process_group("nccl")
rank = dist.get_rank()
torch.cuda.set_device(rank)
device = torch.device(f"cuda:{rank}")

@torch.compile(fullgraph=True)
def step(x, w):
    return (x @ w).sum()

x = torch.randn(8, 8, device=device)
w_shard = torch.randn(4, 8, device=device)   # each rank holds 4 rows

# Step 0 traces with ACT (the native all-gather output); step 1 provides a
# resolved Tensor. This is the safe-to-relax direction (ACT at trace -> Tensor
# at runtime), which the runtime wrapper already handles by unwrapping ACT.
for i in range(6):
    w = all_gather_tensor(w_shard, gather_dim=0, group=dist.group.WORLD)
    if i % 2 == 1:
        # resolve to a plain Tensor before the compiled region
        w = torch.ops._c10d_functional.wait_tensor(w)
    # else (incl. step 0): pass the AsyncCollectiveTensor straight in
    out = step(x, w)
    if rank == 0:
        print(f"step {i}: type(w)={type(w).__name__}, out={out.item():.4f}")

dist.destroy_process_group()
```

Run with:
```
TORCH_LOGS=recompiles torchrun --nproc_per_node=2 repro.py
```

## Expected

The compiled region treats `AsyncCollectiveTensor` (which is awaited before
execution) and the resolved `Tensor` as equivalent — one compiled graph, no
recompiles after the first.

## Actual

Dynamo emits a recompile each time the tensor class changes. Captured output
(`TORCH_LOGS=recompiles`), here at step 1 when the resolved `Tensor` arrives
after step 0 traced with ACT:
```
TORCH_LOGS=recompiles torchrun --nproc_per_node=2 scripts/repro_async_collective_recompile.py
W0611 15:02:26.326000 2128124 site-packages/torch/distributed/run.py:852] 
W0611 15:02:26.326000 2128124 site-packages/torch/distributed/run.py:852] *****************************************
W0611 15:02:26.326000 2128124 site-packages/torch/distributed/run.py:852] Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
W0611 15:02:26.326000 2128124 site-packages/torch/distributed/run.py:852] *****************************************
NCCL version 2.27.5+cuda12.9
step 0: type(w)=AsyncCollectiveTensor, out=9.8565
[rank0]:V0611 15:02:33.918000 2129572 site-packages/torch/_dynamo/guards.py:4514] [0/1] [__recompiles] Recompiling function step in /home/hosseinkh/pytorch/torchtitan_local/scripts/repro_async_collective_recompile.py:32
[rank0]:V0611 15:02:33.918000 2129572 site-packages/torch/_dynamo/guards.py:4514] [0/1] [__recompiles]     triggered by the following guard failure(s):
[rank0]:V0611 15:02:33.918000 2129572 site-packages/torch/_dynamo/guards.py:4514] [0/1] [__recompiles]     - 0/0: ___check_type_id(w, 114219360), type=<class 'torch.distributed._functional_collectives.AsyncCollectiveTensor'>  # return (x @ w).sum()  # pytorch/torchtitan_local/scripts/repro_async_collective_recompile.py:34 in step
[rank1]:V0611 15:02:33.964000 2129573 site-packages/torch/_dynamo/guards.py:4514] [0/1] [__recompiles] Recompiling function step in /home/hosseinkh/pytorch/torchtitan_local/scripts/repro_async_collective_recompile.py:32
[rank1]:V0611 15:02:33.964000 2129573 site-packages/torch/_dynamo/guards.py:4514] [0/1] [__recompiles]     triggered by the following guard failure(s):
[rank1]:V0611 15:02:33.964000 2129573 site-packages/torch/_dynamo/guards.py:4514] [0/1] [__recompiles]     - 0/0: ___check_type_id(w, 114221808), type=<class 'torch.distributed._functional_collectives.AsyncCollectiveTensor'>  # return (x @ w).sum()  # pytorch/torchtitan_local/scripts/repro_async_collective_recompile.py:34 in step
step 1: type(w)=Tensor, out=9.8565
step 2: type(w)=AsyncCollectiveTensor, out=9.8565
step 3: type(w)=Tensor, out=9.8565
step 4: type(w)=AsyncCollectiveTensor, out=9.8565
step 5: type(w)=Tensor, out=9.8565

Note: `out` is identical every step — the two compiled graphs are equivalent, so the recompile from the tensor-class guard is wasted.
```

Note the output is identical every step (the exact value varies per run since
`w` is randomly initialized, but it is constant *within* a run), confirming the
two compiled graphs are equivalent — the recompile is wasted.

## Impact

In TorchTitan's HuggingFace SFT backend (FSDP + TP + compile), this is one of
several guard sources that pushed recompiles past the default
`cache_size_limit=8`, causing `FailOnRecompileLimitHit` with `fullgraph=True`.
Workaround was raising the limit to 64. The recompiles stabilize after ~3-5
steps once the tensor classes settle.

## Environment

- PyTorch: 2.10.0+cu128
- CUDA: 12.8
- GPU: NVIDIA H100 (×8)

## Workaround

```python
torch._dynamo.config.cache_size_limit = 64
```

## Question

Can the class-level guard treat a runtime `Tensor` as compatible with a graph
traced on `AsyncCollectiveTensor` (the ACT-at-trace -> Tensor-at-runtime
direction)? The runtime wrapper already unwraps ACT, so the resolved tensors
produce an equivalent graph, and this is the direction FSDP + TP actually hits.

The reverse direction (Tensor-at-trace -> ACT-at-runtime) should *not* be
relaxed: per discussion, there are cases where the runtime wrapper does not
await ACT until pytorch/pytorch#186442 lands, so relaxing that direction could
silently use an un-awaited ACT.

### Versions

PyTorch: 2.10.0+cu128
CUDA: 12.8
GPU: NVIDIA H100 (×8)

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @jataylo @azahed98

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #189482
* #190576

Fixes #187073.

Root cause: for a traceable wrapper subclass input, Dynamo installs
class-identity guards (a TYPE_MATCH leaf, TENSOR_SUBCLASS_METADATA_MATCH, and
the pytype check baked into the outer TENSOR_MATCH). When a compiled region's
input alternates between an AsyncCollectiveTensor (ACT) and the resolved plain
Tensor -- as happens under FSDP + TP, where an all-gather output is natively an
ACT but is awaited to a plain Tensor before the compiled region on some steps
-- those guards fail on the class change and force a recompile, even though the
two graphs are equivalent (the AOTAutograd runtime wrapper already unwraps ACT
inputs). Users hit this as recompiles blowing past cache_size_limit.

Fix: an ACT flattens to a single inner tensor `elem` whose metadata mirrors the
wrapper, and the AOTAutograd forward runtime unwrap already tolerates a plain
Tensor where a subclass was traced. So for an ACT input, guard the unwrapped
inner tensor through a new guard-only UnwrapCollectiveTensorSource (which yields
x.elem when x is an ACT and x unchanged otherwise) instead of the ACT-locked
class/metadata guards. The traced graph, grapharg, and AOT calling convention
are unchanged -- only guard installation changes -- so an ACT-traced graph is
reused for the resolved plain Tensor with no recompile. Only the
ACT-at-trace -> Tensor-at-runtime direction is relaxed (the relaxation triggers
only when the traced value is an ACT); the reverse direction stays guarded,
since a Tensor-traced graph may skip the collective await.

Soundness: the only way an ACT-traced graph can differ from a plain-Tensor one
is if the region observes the input's Python class. Rather than enumerate
observation channels, the guard is reinstalled on demand at the sites that fold
the class into a value: TensorVariable.var_getattr("__class__") now attaches a
TypeSource (mirroring BuiltinVariable.call_type, so it self-guards downstream --
this is general for all tensors), and call_isinstance installs a TYPE_MATCH for
an ACT arg only when the check actually discriminates ACT from torch.Tensor
(so isinstance(w, torch.Tensor) still reuses, isinstance(w, ACT) recompiles).
type()-based channels already carry provenance via TypeSource and self-guard.

Alternatives considered: (1) a config flag to gate the relaxation was dropped
because, once observation sites self-guard, the change is sound and does not
need an opt-in. (2) Anchoring the ACT inner tensor's ShapeEnv symbols at the
unwrap source (to make dynamic shapes reuse perfectly too) was prototyped and
abandoned: the symbolic-shape and guard machinery hard-require subclass inner
sources to be AttrSource, so this needs a broader design for polymorphic
subclass inner sources. As a result, under dynamic shapes the ACT and the
resolved plain Tensor land in two separate cache entries: the ShapeEnv derives
the inner dim from a plain `.elem` AttrSource rather than the unwrap source, so
the symbolic-shape guard evaluates `L['w'].elem.size()` and fails with an
AttributeError on a plain Tensor. This split is one-time and stable (each class
always hits its own entry; 4 and 8 reps both compile exactly twice), not a
per-alternation recompile. Static shapes reuse perfectly.

Test Plan:

```
python test/dynamo/test_subclasses.py -k async_collective_tensor
python test/dynamo/test_subclasses.py -k dunder_class_provenance
python test/dynamo/test_subclasses.py
python test/dynamo/test_functions.py
```

Distributed (requires 2 GPUs):

```
python test/distributed/test_inductor_collectives.py -k async_collective_tensor
```

This PR was authored with the assistance of an AI coding assistant.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

File: torch/_dynamo/source.py [hunk 1]
```diff
@@ -1183,6 +1183,27 @@ def _name_template(self) -> str:
         return "{0}.__tensor_flatten__()[0]"
 
 
+# Guard-only source that yields the inner tensor of an AsyncCollectiveTensor
+# (ACT) and the base value unchanged for a plain Tensor. Used to guard on the
+# unwrapped tensor so a graph traced on an ACT can be reused when the resolved
+# plain Tensor is passed at runtime. See
+# torch._dynamo.variables.builder.VariableBuilder.wrap_tensor.
+@dataclass_with_cached_hash(frozen=True)
+class UnwrapCollectiveTensorSource(ChainedSource):
```
Comment (aorenste) [inline torch/_dynamo/source.py:1192]: Anchored here for UI visibility -- see the review body for the two test-coverage findings (nested ACT-in-DTensor reuse; serialized-guard round-trip). This `UnwrapCollectiveTensorSource` is the guard source both findings turn on.

File: torch/_dynamo/variables/builder.py [hunk 2]
```diff
@@ -3080,7 +3081,32 @@ def wrap_tensor(self, value: torch.Tensor) -> VariableTracker:
...
+        # correct (ACT desugars ops to the inner tensor), but a compiled region
+        # can succeed where eager would raise AttributeError on the plain Tensor;
+        # that only bites user code already broken in eager on the resolved input.
+        is_polymorphic_act = False
+        if not is_dtensor and is_traceable_wrapper_subclass(value):
+            from torch._functorch._aot_autograd.utils import (
+                is_async_collective_tensor_type,
+            )
+
+            is_polymorphic_act = is_async_collective_tensor_type(type(value))
```
Comment (aorenste) [inline torch/_dynamo/variables/builder.py]: nit: Unnecessary complexity - I would do the import at top-level and just check `is_async_collective_tensor_type` (which is actually going to be faster to check than `is_traceable_wrapper_subclass` anyway)

File: torch/_dynamo/guards.py [hunk 3]
```diff
@@ -778,6 +779,29 @@ def from_numpy(a: Any) -> torch.Tensor:
         return torch.as_tensor(a) if isinstance(a, (np.generic, np.ndarray)) else a
 
 
+_async_collective_tensor_type: type | None = None
+
+
+def unwrap_async_collective_tensor(x: Any) -> Any:
```
Comment (aorenste) [inline torch/_dynamo/guards.py]: nit: Should be either:
```
def unwrap_async_collective_tensor(x: Tensor) -> Tensor:
```
or possibly:
```
def unwrap_async_collective_tensor(x: T) -> T | Tensor:
```

**Review by aorenste — APPROVED** [review]
(Reviewed by me, assisted by AI)

[cleanup] Only top-level ACT inputs are tested (`fn(x, w)` with `w` an ACT or plain Tensor), but the FSDP+TP shape this PR targets is an ACT nested inside `DTensor._local_tensor` (async all-gather / `redistribute(async_op=True)`) -- the shape #186442 was built for. Tracing the guards shows this case reuses by the same mechanism as the flat one: the recursion into `_local_tensor` re-enters `wrap_tensor`, hits `is_polymorphic_act`, and relaxes the inner guard via `UnwrapCollectiveTensorSource(AttrSource(dt, "_local_tensor"))`, while the outer DTensor guards (`TYPE_MATCH`, `DTENSOR_SPEC_MATCH`, `requires_grad`) don't discriminate on the local's class. So it should reuse -- but there's no regression test locking it in, even though it is the motivating scenario. Add one: it can run single-process on a fake PG in `TestDTensorCompile` (the Dynamo twin of `test_direct_aot_dtensor_local_tensor_act_to_plain_no_crash`), no multi-GPU needed.

[cleanup] My agent says: The serialization coverage for this relaxation stops one assertion short of its core property. `test_transparent_subclass_tensor_match` (added in the stacked #190576) round-trips an `AsyncCollectiveTensor` input, and at this PR's HEAD it does exercise the new `UnwrapCollectiveTensorSource` guard path -- but it only asserts the reloaded guard matches an ACT (`_test_check_fn(ref, loaded, {"w": AsyncCollectiveTensor(base)}, True)`). It never asserts the reloaded guard also matches the resolved plain `Tensor` (`_test_check_fn(ref, loaded, {"w": base}, True)`) -- which is exactly the ACT->Tensor reuse this PR exists to enable, and the case that matters for precompile under FSDP+TP. The property holds by construction (guard state is rebuilt by name on load rather than pickled), so this is a lock-in regression test, not a suspected bug -- low priority. The assertion lands in #190576's file (`test/dynamo/test_guard_serialization.py`), not this one.
